File: W3/homeWork_W3/С++/Unit/Unit.cpp
#include "Unit.h"

using namespace std;

void Unit::ensureIsAlive() {
    if ( this->hitPoints == 0 ) {
        throw UnitIsDead();
    }
}

Unit::Unit(const string& name, int hp, int dmg) {
    if ( hp < 0 ) {
        hp = 0;
    }
    if ( dmg < 0 ) {
        dmg = 0;
    }
    this->name = name;
    this->hitPointsLimit = hp;
    this->hitPoints = hp;
    this->damage = dmg;
}

Unit::~Unit() {
    
}

int Unit::getDamage() const {
    return this->damage;
}

int Unit::getHitPoints() const {
    return this->hitPoints;
}

int Unit::getHitPointsLimit() const {
    return this->hitPointsLimit;
}

const string& Unit::getName() const {
    return this->name;
}

void Unit::addHitPoints(int hp) {
    this->ensureIsAlive();
    
    if ( hp < 0 ) {
        hp = 0;
    }
    if ( this->hitPoints + hp > this->hitPointsLimit ) {
        this->hitPoints = this->hitPointsLimit;
        return;
    }
    this->hitPoints += hp;
}

void Unit::takeDamage(int dmg) {
    this->ensureIsAlive();
    
    if ( dmg < 0 ) {
        dmg = 0;
    }
    if ( this->hitPoints < dmg ) {
        this->hitPoints = 0;
        return;
    }
    this->hitPoints -= dmg;
}

void Unit::attack(Unit& enemy) {
    this->ensureIsAlive();
    enemy.takeDamage(this->damage);
    enemy.counterAttack(*this);
}

void Unit::counterAttack(Unit& enemy) {
    this->ensureIsAlive();
    enemy.takeDamage(this->damage / 2);
}

ostream& operator<<(ostream& out, const Unit& unit) {
    out << "Characteristics of the warrior:\n"
        << "Name - " << unit.getName() << ".\n"
        << "Max hit points - " << unit.getHitPointsLimit() << ".\n"
        << "Hit points - " << unit.getHitPoints() << ".\n"
        << "Damage - " << unit.getDamage() << ".";
        
    return out;
}

File: W3/homeWork_W3/Java/Unit/Unit.java
public class Unit {
    private int damage;
    private int hitPoints;
    private int hitPointsLimit;
    private String name;
    
    private void ensureIsAlive() throws UnitIsDeadException {
        if ( this.hitPoints == 0 ) {
            throw new UnitIsDeadException();
        }
    }

    public Unit(String name, int hp, int dmg) {
        if ( hp < 0 ) {
            hp = 0;
        }
        if ( dmg < 0 ) {
            dmg = 0;
        }
        this.name = name;
        this.hitPointsLimit = hp;
        this.hitPoints = hp;
        this.damage = dmg;
    }

    public Unit() {
        this("Knight", 100, 100);
    }

    public final int getDamage() {
        return this.damage;
    }

    public final int getHitPoints() {
        return this.hitPoints;
    }

    public final int getHitPointsLimit() {
        return this.hitPointsLimit;
    }

    public final String getName() {
        return this.name;
    }

    public void addHitPoints(int hp) throws UnitIsDeadException {
        this.ensureIsAlive();
        
        if ( hp < 0 ) {
            hp = 0;
        }
        if ( this.hitPoints + hp > this.hitPointsLimit ) {
            this.hitPoints = this.hitPointsLimit;
            return;
        }
        this.hitPoints += hp;
    }
    
    public void takeDamage(int dmg) throws UnitIsDeadException {
        this.ensureIsAlive();
        
        if ( dmg < 0 ) {
            dmg = 0;
        }
        if ( this.hitPoints < dmg ) {
            this.hitPoints = 0;
            return;
        }
        this.hitPoints -= dmg;
    }
    
    public void attack(Unit enemy) throws UnitIsDeadException {
        this.ensureIsAlive();
        enemy.takeDamage(this.damage);
        enemy.counterAttack(this);
    }
    
    public void counterAttack(Unit enemy) throws UnitIsDeadException {
        this.ensureIsAlive();
        enemy.takeDamage(this.damage / 2);
    }
    
    @Override
    public String toString() {
        return "Characteristics of the warrior:\n"
            + "Name - " + this.name + ".\n"
            + "Max hit points - " + this.hitPointsLimit + ".\n"
            + "Hit points - " + this.hitPoints + ".\n"
            + "Damage - " + this.damage + ".";
    }
}
